libvast/builtins/operators/to.cpp

```diff
@@ -0,0 +1,119 @@
+      parsers::optional_ws_or_comment, parsers::extractor, parsers::identifier,
+      parsers::extractor_char, parsers::extractor_list;
+    const auto* f = pipeline.begin();
+    const auto* const l = pipeline.end();
+    const auto p = optional_ws_or_comment >> identifier
+                   >> -(required_ws_or_comment >> string_parser{"write"}
+                        >> required_ws_or_comment >> identifier)
+                   >> optional_ws_or_comment >> end_of_pipeline_operator;
+    auto result = std::tuple{
+      std::string{}, std::optional<std::tuple<std::string, std::string>>{}};
+    if (!p(f, l, result)) {
+      return {
+        std::string_view{f, l},
+        caf::make_error(ec::syntax_error, fmt::format("failed to parse "
+                                                      "to operator: '{}'",
+                                                      pipeline)),
+      };
+    }
+    auto dumper_name = std::get<0>(result);
+    auto dumper = plugins::find<dumper_plugin>(dumper_name);
+    if (!dumper) {
+      return {std::string_view{f, l},
+              caf::make_error(ec::syntax_error,
+                              fmt::format("failed to parse "
+                                          "write operator: no '{}' dumper "
+                                          "found",
+                                          dumper_name))};
+    }
+    const printer_plugin* printer;
+    auto printer_argument = std::get<1>(result);
+    if (printer_argument) {
+      auto printer_name = std::get<1>(*printer_argument);
+      printer = plugins::find<printer_plugin>(printer_name);
+      if (!printer) {
+        return {std::string_view{f, l},
+                caf::make_error(ec::syntax_error,
+                                fmt::format("failed to parse "
+                                            "to operator: no '{}' printer "
+                                            "found",
+                                            printer_name))};
+      }
+    } else {
+      printer = dumper->make_default_printer();
+    }
+    if (dumper->dumper_requires_joining()
+        and not printer->printer_allows_joining()) {
+      return {std::string_view{f, l},
+              caf::make_error(ec::invalid_argument,
+                              fmt::format("failed to parse "
+                                          "to operator: printer '{}' does not "
+                                          "allow joining but dumper '{}' "
+                                          "requires joining",
+                                          printer->name(), dumper->name()))};
+    }
+    auto print_op = std::make_unique<print_operator>(std::move(*printer));
+    auto dump_op = std::make_unique<dump_operator>(std::move(*dumper));
+    auto ops = std::vector<logical_operator_ptr>{};
+    ops.emplace_back(std::move(print_op));
+    ops.emplace_back(std::move(dump_op));
+    auto sub_pipeline = logical_pipeline::make(std::move(ops));
```

Comment:
IIUC this does not correctly handle the two cases where the dumper does not require its inputs to be joined. We don't currently have such a dumper, but at least for now we should probably error in these cases or implement it correctly.

Reply:
I would say yes. We probably need to instantiate the dump operator once per schema in this case.

Reply:
I implemented a `print_dump_operator` in https://github.com/tenzir/vast/pull/3024/commits/a8cbd8ca23a07c77105893e00786e7db40e7a05f